{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "K & D = 4 2\n",
      "Patterns = ['ACAC|CTCT', 'ACAT|CTCA', 'CACA|TCTC', 'GACA|TCTC']\n",
      "GACACATTCTCA\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "import itertools\n",
    "\n",
    "def debruijn(kmers):\n",
    "    result = {}\n",
    "    for pattern in kmers:\n",
    "        result.setdefault(pattern[:-1], [])\n",
    "        result[pattern[:-1]].append(pattern[1:])\n",
    "    return result\n",
    "\n",
    "def calculate_degree(graph):\n",
    "    degree = {}\n",
    "    for node in graph:\n",
    "        degree.setdefault(node, [0,0]) #[indegree, outdegree]\n",
    "        degree[node][1] += len(graph[node])\n",
    "        for target_node in graph[node]:\n",
    "            degree.setdefault(target_node, [0,0])\n",
    "            degree[target_node][0] += 1\n",
    "    return degree\n",
    "\n",
    "def calculate_start_node(degree):\n",
    "    return [key for key in degree if degree[key][0] < degree[key][1]][0]\n",
    "\n",
    "def eulerian_path(graph):\n",
    "    stack = []\n",
    "    path = []\n",
    "    degree = calculate_degree(graph)\n",
    "    current_node = calculate_start_node(degree)\n",
    "    while True:\n",
    "        if current_node in graph and len(graph[current_node]) > 0:\n",
    "            stack.append(current_node)\n",
    "            target_node = graph[current_node][0]\n",
    "            graph[current_node].remove(target_node)\n",
    "            current_node = target_node\n",
    "        else:\n",
    "            path.insert(0, current_node)\n",
    "            if len(stack) > 0:\n",
    "                current_node = stack.pop()\n",
    "            else:\n",
    "                break\n",
    "    return path\n",
    "\n",
    "def construct_string(path):\n",
    "    string = path[0][0]\n",
    "    for p in path[1:]:\n",
    "        #print(p)\n",
    "        string += p[0][-1]\n",
    "    for p in path[-k-d:]:\n",
    "        string += p[1][-1]\n",
    "    return string\n",
    "\n",
    "def de_bruijn_from_paired_reads(patterns, k):\n",
    "    graph = {}\n",
    "    for text in patterns:\n",
    "        prefix = (text[:k-1], text[k+1:-1])\n",
    "        suffix = (text[1:k], text[k+2:])\n",
    "        if prefix not in graph.keys():\n",
    "            graph[prefix] = []\n",
    "        graph[prefix].append(suffix)\n",
    "    path = eulerian_path(graph)\n",
    "    #print(path)\n",
    "    string = construct_string(path)\n",
    "    return string\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    k = 4\n",
    "    d = 2\n",
    "\n",
    "    # paired_string ='GAGA|TTGA TCGT|GATG CGTG|ATGT TGGT|TGAG GTGA|TGTT GTGG|GTGA TGAG|GTTG GGTC|GAGA GTCG|AGAT' # Expected Output = GTGGTCGTGAGATGTTGA\n",
    "    paired_string ='ACAC|CTCT ACAT|CTCA CACA|TCTC GACA|TCTC' # Expected Output = GACACATCTCTCA\n",
    "    patterns = paired_string.split() # Converts to List, like ['ACAC|CTCT', 'ACAT|CTCA', 'CACA|TCTC', 'GACA|TCTC']\n",
    "\n",
    "    with open(\"input_reconstruct.txt\") as f:\n",
    "        numbers = f.readline().split()\n",
    "        k = int(numbers[0])\n",
    "        d = int(numbers[1])\n",
    "        patterns = f.readlines()\n",
    "        #patterns = [p.strip() for p in patterns]\n",
    "    patterns = patterns[0].split()\n",
    "\n",
    "    print('K & D =', k, d)\n",
    "    print('Patterns =', patterns)\n",
    "    result = de_bruijn_from_paired_reads(patterns, k)\n",
    "    print(result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG ATG TGT TGGA CAT GAT AGA "
     ]
    }
   ],
   "source": [
    "def debruijn(kmers):\n",
    "    result = {}\n",
    "    for pattern in kmers:\n",
    "        result.setdefault(pattern[:-1], [])\n",
    "        result[pattern[:-1]].append(pattern[1:])\n",
    "    return result\n",
    "\n",
    "def calculate_degree(graph):\n",
    "    degree = {}\n",
    "    for node in graph:\n",
    "        degree.setdefault(node, [0,0]) #[indegree, outdegree]\n",
    "        degree[node][1] += len(graph[node])\n",
    "        for target_node in graph[node]:\n",
    "            degree.setdefault(target_node, [0,0])\n",
    "            degree[target_node][0] += 1\n",
    "    return degree\n",
    "\n",
    "def genome_path(patterns):\n",
    "    genome = patterns[0]\n",
    "    for pattern in patterns[1:]:\n",
    "    \tgenome += pattern[-1]\n",
    "    return genome\n",
    "\n",
    "def maximal_non_branching_paths(graph):\n",
    "    paths = []\n",
    "    visited = [] \n",
    "    degree = calculate_degree(graph)\n",
    "    for current_node in graph:\n",
    "        if degree[current_node][0] != 1 or degree[current_node][1] != 1:  \n",
    "            visited.append(current_node)\n",
    "            if degree[current_node][1] > 0:\n",
    "                for target_node in graph[current_node]:\n",
    "                    path = [current_node, target_node]\n",
    "                    while degree[target_node][0] == 1 and degree[target_node][1] == 1:\n",
    "                        visited.append(target_node)\n",
    "                        target_node = graph[target_node][0]\n",
    "                        path.append(target_node)\n",
    "                    paths.append(path)\n",
    "    for current_node in graph:\n",
    "        if degree[current_node][0] == 1 and degree[current_node][1] == 1:\n",
    "            if current_node not in visited:\n",
    "                visited.append(current_node)\n",
    "                target_node = graph[current_node][0]\n",
    "                cycle = [current_node]\n",
    "                while degree[target_node][0] == 1 and degree[target_node][1] == 1:\n",
    "                    visited.append(target_node)\n",
    "                    cycle.append(target_node)\n",
    "                    if current_node == target_node:\n",
    "                        break\n",
    "                    target_node = graph[target_node][0]\n",
    "                paths.append(cycle)\n",
    "    return paths\n",
    "\n",
    "def ContigsFromReads(patterns):\n",
    "    contigs = []\n",
    "    graph = debruijn(patterns)\n",
    "    #print(graph)\n",
    "    paths = maximal_non_branching_paths(graph)\n",
    "    #print(paths)\n",
    "    for path in paths:\n",
    "        contigs.append(genome_path(path))\n",
    "    return contigs\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    patterns = ['ATG', 'ATG', 'TGT', 'TGG', 'CAT', 'GGA', 'GAT', 'AGA']\n",
    "    # with open(\"rosalind_ba3k.txt\") as f:\n",
    "    #     f = f.read().splitlines()\n",
    "    #     patterns = []\n",
    "    #     for i in range(len(f)):\n",
    "    #         patterns.append(f[i])\n",
    "    #     #print(patterns)\n",
    "    contigs = ContigsFromReads(patterns)\n",
    "    out = open(\"out_ba3k.txt\", \"w\")\n",
    "    #out.writelines(contigs)\n",
    "    #out.close()\n",
    "    for contig in contigs:\n",
    "        out.write(contig)\n",
    "        out.write(\"\\n\")\n",
    "        print(contig, end=\" \")\n",
    "    out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGAG GACTG CTGA CTGA\n"
     ]
    }
   ],
   "source": [
    "kmer_string = 'TGAG GACT CTGA ACTG CTGA'\n",
    "patterns = kmer_string.split()\n",
    "\n",
    "# patterns = ['ATG', 'ATG', 'TGT', 'TGG', 'CAT', 'GGA', 'GAT', 'AGA']\n",
    "contigs = ContigsFromReads(patterns)\n",
    "\n",
    "print(' '.join(str(e) for e in contigs))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
