{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Methods from Previous Weeks\n",
    "\n",
    "def Neighbors(pattern, d):\n",
    "    bases     = ['A', 'C', 'G', 'T']\n",
    "    \n",
    "    if d == 0:\n",
    "        return (pattern)\n",
    "    if len(pattern) == 1:\n",
    "        return (bases)\n",
    "\n",
    "    neighbors       = []\n",
    "    suffix_neighbors = Neighbors(pattern[1:], d)\n",
    "    for text in suffix_neighbors:\n",
    "        if HammingDistance(pattern[1:], text) < d:\n",
    "            for base in bases:\n",
    "                neighbors.append(base + text)\n",
    "        else :\n",
    "            neighbors.append(pattern[0] + text)\n",
    "    return(neighbors)\n",
    "\n",
    "def HammingDistance(seq1, seq2):\n",
    "    if len(seq1) != len(seq2):\n",
    "        raise ValueError(\"Input sequences must have equal length.\")\n",
    "    \n",
    "    distance = 0\n",
    "    for base1, base2 in zip(seq1, seq2):\n",
    "        if base1 != base2:\n",
    "            distance += 1\n",
    "\n",
    "    return distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Number of Strings = 496000\n",
      "Expected Number of Occurences = 1.89208984375\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.2 - Exercise Break\n",
    "\n",
    "# What is the expected number of occurrences of a 9-mer in 500 random DNA strings, \n",
    "# each of length 1000? Assume that the sequences are formed by selecting each \n",
    "# nucleotide (A, C, G, T) with the same probability (0.25).\n",
    "\n",
    "# Note: Express your answer as a decimal; allowable error = 0.0001.\n",
    "\n",
    "string_count = 500\n",
    "length = 1000\n",
    "k = 9\n",
    "nucleotide_probability = .25\n",
    "\n",
    "total_strings = string_count * (1 + (length - k))\n",
    "\n",
    "num_of_mismatches = total_strings * (nucleotide_probability ** k)\n",
    "\n",
    "print(\"Total Number of Strings =\", total_strings)\n",
    "print(\"Expected Number of Occurences =\", num_of_mismatches)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample Implanted Motifs = GTT ATT ATA TTT\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.2 - Mofif Problem\n",
    "\n",
    "# Given a collection of strings Dna and an integer d, a k-mer is a (k,d)-motif \n",
    "# if it appears in every string from Dna with at most d mismatches. For example, \n",
    "# the implanted 15-mer in the strings above represents a (15,4)-motif.\n",
    "\n",
    "# Implanted Motif Problem: Find all (k, d)-motifs in a collection of strings.\n",
    "\n",
    "# Input: A collection of strings Dna, and integers k and d.\n",
    "# Output: All (k, d)-motifs in Dna.\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "def MotifEnumeration(genome, k, d):\n",
    "    genome = genome.split(\" \") \n",
    "    patterns     = []\n",
    "    patterns_all = []\n",
    "    for text in genome:\n",
    "        patterns_text = []\n",
    "        for start in range(len(text) - k + 1):\n",
    "            end     = start + k\n",
    "            pattern = list(text[start: end])\n",
    "            patterns_text.extend(Neighbors(pattern, d)) \n",
    "            \n",
    "        patterns_text = list(set(patterns_text))\n",
    "        patterns_all.extend(patterns_text)\n",
    "        \n",
    "    patterns_count = Counter(patterns_all)\n",
    "    for pattern, count in patterns_count.items():\n",
    "        if count == len(genome):\n",
    "            patterns.append(pattern)\n",
    "            \n",
    "    return (patterns)\n",
    "\n",
    "# Sample Test (output = ATA ATT GTT TTT)\n",
    "k    = 3\n",
    "d    = 1\n",
    "genome ='ATTTGGC TGCCTTA CGGTATC GAAAATT'\n",
    "\n",
    "print(\"Sample Implanted Motifs =\", \" \".join(MotifEnumeration(genome, k, d)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Randomized Implanted Motifs = AGCGT GCGTC GCATC GCCTC GCTTG GCTTC\n"
     ]
    }
   ],
   "source": [
    "# Randomized Problem\n",
    "k    = 5\n",
    "d    = 1\n",
    "genome ='GCATCATCCTGCGTTTGGGGGAAGG TGAGGACTCTGCTACAAGGAGCTTC GCTTCCGAGTAGGGGTTACCGTTAC AGAGTGCTTCTGGCGTCCACAGACG GGCGTCAACATCCTTGACAGGCGTC TGCTCGGGAATTACTAAACAGCGTC'\n",
    "\n",
    "print(\"Randomized Implanted Motifs =\", \" \".join(MotifEnumeration(genome, k, d)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score = 105\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.3 - Scoring Motifs\n",
    "\n",
    "# The minimium possible value of Score(Motifs) is 0 (if all the k-mers in \n",
    "# Motifs are the same). What is the maximum possible value of Score(Motifs) \n",
    "# for 10 motifs of length 15?\n",
    "\n",
    "# Not sure how to program this, but here's my thinking.\n",
    "\n",
    "motifs_count = 10\n",
    "length = 15\n",
    "lower_case_no_color_count = 3\n",
    "\n",
    "print(\"Score =\", (motifs_count - lower_case_no_color_count) * length)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30\n",
      "[[0.2 0.2 0.  0.  0.  0.  0.9 0.1 0.1 0.1 0.3 0. ]\n",
      " [0.1 0.6 0.  0.  0.  0.  0.  0.4 0.1 0.2 0.4 0.6]\n",
      " [0.  0.  1.  1.  0.9 0.9 0.1 0.  0.  0.  0.  0. ]\n",
      " [0.7 0.2 0.  0.  0.1 0.1 0.  0.5 0.8 0.7 0.3 0.4]]\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.3 - Scoring Motifs Continued\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def ProfileMatrix(motifs):\n",
    "    if type(motifs) == str:\n",
    "        strings  = list(motifs)\n",
    "        n_string = len(strings)\n",
    "        k        = len(strings[0])\n",
    "        profile_matrix  = np.array([0] * 4 * k, dtype='f').reshape(4, k)\n",
    "    \n",
    "    if type(motifs) == list:\n",
    "        motifs_array = np.array([list(seq) for seq in motifs])\n",
    "        profile_matrix = np.zeros((4, motifs_array.shape[1]))\n",
    "\n",
    "    for i in range(motifs_array.shape[1]):\n",
    "        column_i = motifs_array[:, i]\n",
    "        len_column_i = column_i.shape[0]\n",
    "        count = Counter(column_i)\n",
    "        profile_matrix[0, i] = count[\"A\"] / len_column_i\n",
    "        profile_matrix[1, i] = count[\"C\"] / len_column_i\n",
    "        profile_matrix[2, i] = count[\"G\"] / len_column_i\n",
    "        profile_matrix[3, i] = count[\"T\"] / len_column_i\n",
    "\n",
    "    return profile_matrix\n",
    "\n",
    "\n",
    "def Score(motifs):\n",
    "    motifs_array = np.array([list(seq) for seq in motifs])\n",
    "    score = 0\n",
    "\n",
    "    for i in range(motifs_array.shape[1]):\n",
    "        column_i = motifs_array[:, i]\n",
    "        count = Counter(column_i)\n",
    "        max_freq = count.most_common(1)[0][1]\n",
    "        score += column_i.shape[0] - max_freq\n",
    "\n",
    "    return score\n",
    "\n",
    "\n",
    "motifs = [\n",
    "    \"TCGGGGGTTTTT\",\n",
    "    \"CCGGTGACTTAC\",\n",
    "    \"ACGGGGATTTTC\",\n",
    "    \"TTGGGGACTTTT\",\n",
    "    \"AAGGGGACTTCC\",\n",
    "    \"TTGGGGACTTCC\",\n",
    "    \"TCGGGGATTCAT\",\n",
    "    \"TCGGGGATTCCT\",\n",
    "    \"TAGGGGAACTAC\",\n",
    "    \"TCGGGTATAACC\",\n",
    "]\n",
    "\n",
    "print(Score(motifs))\n",
    "print(ProfileMatrix(motifs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "[[0.2 0.8 0.4 0.2 0.4 0.4 0.2 0.8 0.  0.  0.2 0.4]\n",
      " [0.6 0.  0.4 0.  0.  0.2 0.4 0.  0.  0.4 0.6 0.4]\n",
      " [0.2 0.2 0.2 0.6 0.2 0.  0.2 0.  0.4 0.2 0.2 0.2]\n",
      " [0.  0.  0.  0.2 0.4 0.4 0.2 0.2 0.6 0.4 0.  0. ]]\n"
     ]
    }
   ],
   "source": [
    "motifs = [\n",
    "    \"GGCGTTCAGGCA\",\n",
    "    \"AAGAATCAGTCA\",\n",
    "    \"CAAGGAGTTCGC\",\n",
    "    \"CACGTCAATCAC\",\n",
    "    \"CAATAATATTCG\",\n",
    "]\n",
    "\n",
    "print(Score(motifs))\n",
    "print(ProfileMatrix(motifs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 10 motifs of length 12\n",
      "Entropy Score = 9.916290005356972\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.3 - Exercise Break\n",
    "\n",
    "# Compute the entropy of the NF-ÎºB motif matrix (reproduced below).\n",
    "\n",
    "def Entropy(motifs):\n",
    "    Profile = {}\n",
    "\n",
    "    # Check motifs length\n",
    "    list_len = len(motifs)\n",
    "    L1 = len(motifs[1]) # length of the first motif\n",
    "    print('There are {} motifs of length {}'.format(list_len, L1))\n",
    "    \n",
    "    for i in range(len(motifs)):\n",
    "        if len(motifs[i]) != L1:\n",
    "            short_motif = motifs[i]\n",
    "            short_motif_len = len(short_motif)\n",
    "            print('Oops, Motif {} is {} nucleotides instead of {}!'.format(short_motif, short_motif_len, L1))\n",
    "            break\n",
    "        \n",
    "    # Fill positions with frequency of 0\n",
    "    for nucleotide in 'ACGT':\n",
    "        values = [0] * L1\n",
    "        Profile[nucleotide] = values\n",
    "        \n",
    "    # Iterate through each position, counting nucleotide frequencies\n",
    "    total_entropy = 0\n",
    "    for key, values in Profile.items():\n",
    "        for motif in motifs:\n",
    "            for i in range(len(motif)):\n",
    "                if motif[i] == key:\n",
    "                    Profile[key][i] += 1\n",
    "        \n",
    "        # Convert frequencies to probabilities\n",
    "        for i in range(len(values)):\n",
    "            Profile[key][i] = Profile[key][i] / float(list_len)\n",
    "        \n",
    "        # Calculate total entropy (Sum of (Prob_value * log2 Prob_n))\n",
    "        import math\n",
    "        for value in values:\n",
    "            if value > 0:\n",
    "                total_entropy += abs(value * math.log(value, 2))\n",
    "            else: continue\n",
    "            \n",
    "    return(total_entropy)\n",
    "\n",
    "motifs = [\n",
    "\"TCGGGGGTTTTT\",\n",
    "\"CCGGTGACTTAC\",\n",
    "\"ACGGGGATTTTC\",\n",
    "\"TTGGGGACTTTT\",\n",
    "\"AAGGGGACTTCC\",\n",
    "\"TTGGGGACTTCC\",\n",
    "\"TCGGGGATTCAT\",\n",
    "\"TCGGGGATTCCT\",\n",
    "\"TAGGGGAACTAC\",\n",
    "\"TCGGGTATAACC\"\n",
    "]\n",
    "\n",
    "print(\"Entropy Score =\", Entropy(motifs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample Median String = GAC\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.4 Implement MedianString\n",
    "\n",
    "# Input: An integer k, followed by a space-separated collection of strings Dna.\n",
    "# Output: A k-mer Pattern that minimizes d(Pattern, Dna) among all possible choices \n",
    "# of k-mers. (If there are multiple such strings Pattern, then you may return any one.)\n",
    "\n",
    "def Number2Symbol(index):\n",
    "    nucleotide = {\"A\":0, \"C\":1, \"G\":2, \"T\":3}\n",
    "    for k, v in nucleotide.items():\n",
    "        if v == index:\n",
    "            return k\n",
    "        \n",
    "def Number2Pattern(index, k):\n",
    "    if k == 1:\n",
    "        return Number2Symbol(index)\n",
    "    prefix_index = index // 4\n",
    "    r = index % 4 \n",
    "    symbol = Number2Symbol(r)\n",
    "    prefix_pattern = Number2Pattern(prefix_index, k-1)\n",
    "    return \"\".join([prefix_pattern, symbol]) \n",
    "\n",
    "def DistanceBetweenStrings(pattern, dnas):\n",
    "    if type(dnas) == str:\n",
    "        dnas = dnas.split(\" \")\n",
    "        \n",
    "    distance = 0\n",
    "    k        = len(pattern) \n",
    "    \n",
    "    for text in dnas:\n",
    "        H_distance = float('Inf')\n",
    "        for start in range(len(text) - k + 1):\n",
    "            end    = start + k\n",
    "            string = text[start: end]\n",
    "            \n",
    "            if H_distance > HammingDistance(pattern, string):\n",
    "                H_distance = HammingDistance(pattern, string)\n",
    "        distance = distance + H_distance\n",
    "    \n",
    "    return distance\n",
    "\n",
    "def MedianString(genome,k):\n",
    "    \n",
    "     # If genome string, change to list\n",
    "    if type(genome) == str:\n",
    "        genome = genome.split(\" \")\n",
    "    \n",
    "    distance = float('Inf')\n",
    "    median = ''\n",
    "\n",
    "    # Iterate through string patterns\n",
    "    for i in range(4**k):\n",
    "        pattern = Number2Pattern(i, k)\n",
    "        cur_distance = DistanceBetweenStrings(pattern, genome)\n",
    "\n",
    "        if cur_distance < distance:\n",
    "            distance = DistanceBetweenStrings(pattern, genome)\n",
    "            median = pattern\n",
    "    \n",
    "    return median\n",
    "\n",
    "# Sample Test (output = GAC)\n",
    "k    = 3\n",
    "genome ='AAATTGACGCAT GACGACCACGTT CGTCAGCGCCTG GCTGAGCACCGG AGTTCGGGACAG'\n",
    "\n",
    "print(\"Sample Median String =\", MedianString(genome, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample Distance Result = 5\n",
      "Random Distance Result = 64\n"
     ]
    }
   ],
   "source": [
    "# Sample Test (output = 5)\n",
    "pattern = 'AAA'\n",
    "dnas = 'TTACCTTAAC GATATCTGTC ACGGCGTTCG CCCTAAAGAG CGTCAGAGGT'\n",
    "\n",
    "print(\"Sample Distance Result =\", DistanceBetweenStrings(pattern, dnas))\n",
    "\n",
    "# Random Test\n",
    "pattern = 'ACTAAGA'\n",
    "dnas = 'TACTATAGTGGGTACGGTGGTTAGGGACTGATTAGGCGGGTGATGCCAACCGCCTGCCCCTTCACTCTAAGAGCGTAGCAGATAGCTGGGGACTGGTC ACTATTGCTATTACTCTAGAGAAAAATTATCTGCACTATAGCTAGGCCGCAAATCGCTACGAGAAACACCCTCGTGAGCAGACGTGGTATTAAGCTCC GACCAAACCCTCGCCAAAACGATCACGCGGTCTCCTGGTTTCGATGTCGCGCACTTGTAAAGGGGAGTTAGATGTAAACAAGGCTAGCAGTCGTGTTC CTTCGCTGAGTTGGCCGCTGGTGGACGACAACCAACTTGGGTCAAGTTGGGACCAACCGCGCAGCGATGAATGAGCGAAAATCTGTTAAGCGCCGCGG TCGTCCCAAGAATATTGGGCCCCATGCTGTAGAAGTAGATGCATAGACGCTGGCCAGTCTCAGCATGACTCACGACATCCTCGGTTGCGAGCAGAAGG GAGCGATATCGCTCCGTCATTTGCGATTGATTCCCAGTACAACGGGTGGTCTTGGACGTACCTAGTAATACAGAGCACGTCTTGTTGCGAATATAATC GTTACGCATTAGACCGGCTATCTTCCAGCAGATGCTGGGCCCTCTGGTGTTAAGGATCACTCTAAGATGCCTAGCGGATTAGGCTGAAAGAACAACCG CAGACTCACCGGGTAACCGAGTTTGAACGGCAATTACACAGAATATCTGGGCCGAAGTGTGGCGGGGACGGACCAGTTCGCTGCCTACTCACTTTCTA CCTCTGTGAGGTGTATCACAGCAGATGGAAAGTAGCGATATAAGCTAGAAATCAGACCATGCGTCCAAGGGAACACACGGACCTGACCTATTGTTCTC ACTTGCGGGCAGATGCCGATTATAATAAGTGCATGTTTCCTTGCTAGCCGTACGTCAGCATCTCAAATTTGATGATCGCACATAGTGTTCCCTTCACT TAAATGATGAATTGTAATGCTCGGGAACCACGTGTCCTACGAGCTAATAGATTAAGACTAACTTACCGCGCAAGAGATGTATTATCGTAGTCCTCCGT CTACACTCTTGAATTCCTCGCTGGTAAAGGCTACTGTGTAGCGACGTGTTCTTACGTACAGGTGTGCTCATAATATAAGAGCTGCTAGAAGAGAGTAT CAACAAAACGTCGATACGATGTTTGGATGTATATACGACGTTACGTACATGTGTAATGCTCGTCGCCTTAGTATAGCGCGGTAGTGAGGAGAGTCCAT AGAAGGAAGCCTGTTTTCGTCGGTCCCTACGCGTCGAAGGCGCTCTTTCGGATCTTGGCAAAAACAAGCCCCGGAAGACCCAGGAGCATTGTTTGTAA CGCGTGGAATTTTCGCGTTCGTCCCATGTTTGTTAAAAGGACTTGATTGTTAAACCTCGGACTAGCATTCCAGGTTCCGGTAGGGTACGCGCTATAGT ACTATGACGACGTTCAGGTCTTTCAAGAGTGTGAGGAAGAGAATACTGTTCAAGCTTATTTCGCCTAGCCCGGCCAGTCACAGATTGCACGGTTGAGA TCCGAGATCATTCTCGCATAGGCGGACTTATGCTTCACAAGATGCAGGAATCGGAGTCACTTACCTCATTTCTGGTTGAAGGGGAATTTTATATATTG AAATCTTTGGCCCGCTTACCGGTGGCTCAAAATGCCAGCGCGGGAATGGTGCGGATGGTCCTAGACGATTCTCGTCGCGAGTCACGTACATACTTTCT TACCGATATACCTAGGCCCGTATGCTATCGAGTGAGTTTGGCCTTTGTGAGCCTTCGTAGGAAGCGTCATGTCGCAATCACGGATTAATGGCCACAAT GATAGCTCGGAGAGACGTTTCGCTCTTACAAGATCCTTTCTAAAAGAGTTGTGATCTTGCAAAACCTAGTCAACAGGGCAATGGGCGATCCAGGGAGC CAAGCCGGATAACGTTAAACGCAGGGGTATCCCGACCTGGAGCGTAGACAGCGATGAAGCATAGGGATTTCGTGGACTTGCGTGTTTAACCGACCATC AATTCTAGATACACTCTCCCCAAGCTAGATTTACCGTGGAACGCAGCACACTGAATAGGAGACTTCCTCTTGTCCTACTACCCCACCGATTATCCATT TGATTATTGCTCGTGCTAGCCTGAAGAGTCGTCTAGGGACCATTGATCTGGACTATTCGAGCTCGGATCATGCTTTTCATATTCTCATTTCTGCTAAC CGTTTGATGCGTCTATCAACGTCGCCCGAAACTCTTTATCGTAGACATCTGGAGAGATAGTACGAAGTGTAGTCGACTTCGTGCCATGGTATAGAGGT CCGGTACAATGAAGGAACTCTTGTACAAATGAACAGCATGGCGGCGTTACATAGTTGACAGGACTGGATTGCCTTACCAGGGAGCGTCAGTATCGAGA TCCTCTTATGCCCTCCAGAGATGGAATAACGCTAAGGCTATTTTGACGAAACCCCTTAGCGCTGCTCCTCGCTAGCGGTGGCTTTTGGTATCTGGTTC ACATACTGAAATCTTCCAGAATCCTTCTGTGAAAGAAAAGACGAGAGGGGTGAAGCATATTAAGGGGGCCATCTGCCTACGTGAGGATCGCAGGGAGC CACGCGAGGACTCCGGCGAACATAATGCAACATCCTCGTCGGGGCTTACGTCAACGAGAGGTAAACGCTAGGAGAGATATTGATTCGACCGGTCCTCC CCGAATCCGTAAACGTCTATGCAAAAATCCGTCAAGTAGGACTCGAACCAGCGAGACCTAGTTTCTCCCTAGGTCGCCTACGCGCAGGCGGTCTCCAC GAGTGCCTGTATATCGGCCATCGTTATAGGTGATCTCAGTATAACTCGGACCTTGTTTCCATTTTACGTCTTCCTGTCAAAACCAGGCGCCTAAAATC'\n",
    "\n",
    "print(\"Random Distance Result =\", DistanceBetweenStrings(pattern, dnas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Randomized Median String = TTATGG\n"
     ]
    }
   ],
   "source": [
    "# Randomized Test \n",
    "k    = 6\n",
    "genome ='AGCTATTAAATGGCGAAGCTATCTATTCTCTTATGGATATGT CAGTTATCCGCATTTCACTGGGAGGATCTTTTATGGTCTGGT GTTCATCAGGAGATCAGCCAAGTCCCGAGGTGATGGGTACAC CTTAGTCTATCAAACACGGTAGGGTAATGGTTGCATAACCTA GAAGAGCGGTGTGATCTATATGCACGGTAGTTTCGTTTATGG TGCCAACTGCGATAATGGCGCAGACGCTTCTTAACTCCCCGC GCCTAACCGGCACATGTATTATGGCGACAACGTGTGAGGGGA ACGGCACTCATGGGGCAACGGCGGGAGTCATCATGGTTAAAG AACACAGTTAAACGACCGCATGACATTTGATTATGGTCCACG GCGCGCGGAGCAAAGTAATTGGTGTTATGGCGAGATTGATGG'\n",
    "\n",
    "print(\"Randomized Median String =\", MedianString(genome, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample should be 0.0205753ish = 0.020575296\n",
      "Exercise Compute Pr = 0.0\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.5 - Compute Pr(text, profile)\n",
    "\n",
    "# Input:  String text and matrix dictionary profile \n",
    "# Output: Float compute value\n",
    "\n",
    "def ComputePr(text, profile):\n",
    "    p = 1.0000\n",
    "    for i in range(len(text)):\n",
    "        p = p * profile[text[i]][i]\n",
    "    return p\n",
    "\n",
    "text = 'TCGTGGATTTCC'\n",
    "profile = {\n",
    "    'A': [0.2, 0.2, 0.0, 0.0, 0.0, 0.0, 0.9, 0.1, 0.1, 0.1, 0.3, 0.0],\n",
    "    'C': [0.1, 0.6, 0.0, 0.0, 0.0, 0.0, 0.0, 0.4, 0.1, 0.2, 0.4, 0.6],\n",
    "    'G': [0.0, 0.0, 1.0, 1.0, 0.9, 0.9, 0.1, 0.0, 0.0, 0.0, 0.0, 0.0],\n",
    "    'T': [0.7, 0.2, 0.0, 0.0, 0.1, 0.1, 0.0, 0.5, 0.8, 0.7, 0.3, 0.4]\n",
    "}\n",
    "\n",
    "print('Sample should be 0.0205753ish =', ComputePr('TCGGGGATTTCC', profile))\n",
    "print('Exercise Compute Pr =',ComputePr(text, profile))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample w/ dict should be CCGAG = CCGAG\n",
      "Sample w/ list should be CCGAG = CCGAG\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.5 Solve the Profile-most Probable kmer Problem\n",
    "\n",
    "# Input:  String text, integer k and matrix dictionary profile\n",
    "# Output: String most_probable_kmer\n",
    "\n",
    "def ProfileMostProbableKmer(text, k, profile):\n",
    "    max_prob = -1\n",
    "    most_probable_kmer = \"\"\n",
    "\n",
    "    if type(profile) == dict:\n",
    "        for i in range(len(text) - k + 1):\n",
    "            kmer = text[i:i+k]\n",
    "            probability = 1\n",
    "        \n",
    "            for j in range(k):\n",
    "                nucleotide = kmer[j]\n",
    "                probability *= profile[nucleotide][j]\n",
    "            \n",
    "            if probability > max_prob:\n",
    "                max_prob = probability\n",
    "                most_probable_kmer = kmer\n",
    "    \n",
    "    if type(profile) == list:\n",
    "        nucleotide_to_index = {'A': 0, 'C': 1, 'G': 2, 'T': 3}\n",
    "\n",
    "        for i in range(len(text) - k + 1):\n",
    "            kmer = text[i:i+k]\n",
    "            probability = 1\n",
    "\n",
    "            for j in range(k):\n",
    "                nucleotide = kmer[j]\n",
    "                nucleotide_index = nucleotide_to_index[nucleotide]\n",
    "                probability *= profile[nucleotide_index][j]\n",
    "\n",
    "            if probability > max_prob:\n",
    "                max_prob = probability\n",
    "                most_probable_kmer = kmer\n",
    "\n",
    "    return most_probable_kmer\n",
    "\n",
    "text = 'ACCTGTTTATTGCCTAAGTTCCGAACAAACCCAATATAGCCCGAGGGCCT'\n",
    "k = 5\n",
    "profile = {\n",
    "    'A': [0.2, 0.2, 0.3, 0.2, 0.3],\n",
    "    'C': [0.4, 0.3, 0.1, 0.5, 0.1],\n",
    "    'G': [0.3, 0.3, 0.5, 0.2, 0.4],\n",
    "    'T': [0.1, 0.2, 0.1, 0.1, 0.2]\n",
    "}\n",
    "\n",
    "print('Sample w/ dict should be CCGAG =', ProfileMostProbableKmer(text, k, profile))\n",
    "\n",
    "profile = [\n",
    "    [0.2, 0.2, 0.3, 0.2, 0.3],\n",
    "    [0.4, 0.3, 0.1, 0.5, 0.1],\n",
    "    [0.3, 0.3, 0.5, 0.2, 0.4],\n",
    "    [0.1, 0.2, 0.1, 0.1, 0.2]\n",
    "]\n",
    "\n",
    "print('Sample w/ list should be CCGAG =', ProfileMostProbableKmer(text, k, profile))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample output (CAG CAG CAA CAA CAA)? ['GGC', 'AAG', 'CAA', 'CAC', 'CAA']\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.5 Greedy Motif Search Algorithm\n",
    "\n",
    "# Input:  String dnas, integer k (kmer length) and integer t (sequences)\n",
    "# Output: matrix profile\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "def ProfileMatrix(motifs):\n",
    "    n_seqs  = len(motifs)\n",
    "    k       = len(motifs[0])\n",
    "    \n",
    "    profile_matrix = np.zeros((4, k), dtype='f')\n",
    "\n",
    "    for i in range(k):\n",
    "        base_i = []\n",
    "        for string in motifs:\n",
    "            if len(string) >= k:  # Check sequence length\n",
    "                base_i.append(string[i])\n",
    "        count_base    = Counter(base_i)\n",
    "        profile_matrix[0, i] = count_base['A'] / n_seqs\n",
    "        profile_matrix[1, i] = count_base['C'] / n_seqs\n",
    "        profile_matrix[2, i] = count_base['G'] / n_seqs\n",
    "        profile_matrix[3, i] = count_base['T'] / n_seqs\n",
    "\n",
    "    return profile_matrix\n",
    "\n",
    "def GreedyMotifSearch(sequences, k, t):\n",
    "    best_motifs = [sequence[:k] for sequence in sequences]\n",
    "    best_score = float('inf')\n",
    "\n",
    "    for i in range(len(sequences[0]) - k + 1):\n",
    "        motifs = [sequences[0][i:i + k]]\n",
    "        score = 0.0\n",
    "        profile = []\n",
    "\n",
    "        for j in range(1, t):\n",
    "            profile = ProfileMatrix(motifs)\n",
    "            most_probable_kmer = ProfileMostProbableKmer(sequences[j], k, profile) \n",
    "            motifs.append(most_probable_kmer)\n",
    "\n",
    "        best_score = DistanceBetweenStrings(best_motifs[0], best_motifs[1:])\n",
    "        current_score = DistanceBetweenStrings(motifs[0], motifs[1:])\n",
    "\n",
    "        # Can also use\n",
    "        # best_score    = Score(best_motifs)\n",
    "        # current_score = Score(motifs)\n",
    "\n",
    "        if best_score > current_score:\n",
    "            best_motifs = motifs\n",
    "            \n",
    "    return best_motifs\n",
    "\n",
    "# Sample Test\n",
    "k    = 3 \n",
    "t    = 5\n",
    "dnas = 'GGCGTTCAGGCA AAGAATCAGTCA CAAGGAGTTCGC CACGTCAATCAC CAATAATATTCG'\n",
    "sequences = dnas.split(\" \")\n",
    "\n",
    "print(\"Sample output (CAG CAG CAA CAA CAA)?\", GreedyMotifSearch(sequences, k, t))\n",
    "\n",
    "# TODO: Not returning expected results, Need to investigate online\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample output (CAG CAG CAA CAA CAA)? CAG CAG CAA CAA CAA\n"
     ]
    }
   ],
   "source": [
    "# From https://github.com/andmedina/greedyMotifSearch/blob/main/greedyMotif.py\n",
    "\n",
    "# Input:  A list of kmers Dna, and integers k and t (where t is the number of kmers in Dna)\n",
    "# Output: GreedyMotifSearch(Dna, k, t)\n",
    "\n",
    "def Profile(Motifs):\n",
    "    k = len(Motifs[0])\n",
    "    nuc_freq = {symbol: [0]*k for symbol in \"ACGT\"}\n",
    "    for row in Motifs:\n",
    "        for index, char in enumerate(row):\n",
    "            nuc_freq[char][index] += 1/k    \n",
    "    return nuc_freq\n",
    "\n",
    "def GreedyMotifSearch(Dna, k, t):\n",
    "    if type(Dna) == str:\n",
    "        Dna = Dna.split(\" \")\n",
    "\n",
    "    BestMotifs = []\n",
    "    for i in range(t): #range over strings in Dna\n",
    "        BestMotifs.append(Dna[i][0:k]) # sets best motifs to first 3-mer in each dna string\n",
    "    n = len(Dna[0])\n",
    "    for i in range(n-k+1): #range over length of a string in Dna (all of them are the same length)\n",
    "        Motifs = [] #empty list of Motifs\n",
    "        Motifs.append(Dna[0][i:i+k]) #get Motif from first string in Dna consecutive possible 3-mers\n",
    "        for j in range(1,t): #range over from Dna[2nd string] until the end of Dna strings\n",
    "            P = Profile(Motifs[0:j]) #Get profile of probabilities for Motif previously obtained against the strings below  \n",
    "            Motifs.append(ProfileMostProbableKmer(Dna[j],k,P)) #append most probable k-mers found in Dna strings 2 to 5 to the list based on the motif obtained from Dna string 1\n",
    "        if Score(Motifs) < Score(BestMotifs): # create consensus string, and then compute dissimilarity scores for current Motifs and previous BestMotifs, if the current is less (i.e. better) than the previous. \n",
    "            BestMotifs = Motifs #update the best as the current Motifs.\n",
    "    return BestMotifs\n",
    "\n",
    "# Sample Test\n",
    "k    = 3 \n",
    "t    = 5\n",
    "dnas = 'GGCGTTCAGGCA AAGAATCAGTCA CAAGGAGTTCGC CACGTCAATCAC CAATAATATTCG'\n",
    "\n",
    "print('Sample output (CAG CAG CAA CAA CAA)?', ' '.join(str(e) for e in GreedyMotifSearch(dnas, k, t)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Random Test\n",
    "k    = 12 \n",
    "t    = 25\n",
    "dnas = 'GGCCAGGTTGTCTTCAGATAAATGGAATTGCAAGTGGAGAGGCGGATGCCTGACCACAGGAACACGAGCACCGAACTCCGCCTCATACAGCTTCACACCACACGATATCCATAGGCCGCGTACGGACCGTGGCCTTGGCTAGTCCGCTGCGGGAAG AAAGGTTCCCTGATAACCGAATTACGGAAGGCATGTCTACAGTATAGTTTCAGATATAGTTATTCTAGAAAAAGATGTAGGCCCCTTCAGCGTATTACTCTACACAGTCGGAGTAAGAGTATATAGTTTGGACTATTAATTCAATGTCACGACAAT CTCTCATGTCAAGCTCCGTCGCCATCCTCTGGTTCGCGAATTGCATTGTTGCTACGTAGGCACAGGGCGACGGTCCAGCCGCGCGATCCACGATGCGAGGGGAAATTCCTCAGATAGATGGGCAGGTTATTGGTTCTAGTCCGTTGTGGGACGAGT TAACCTCATAAATTTATTACTCTAGTTGATTCACTCCTATTATTGCTTGTAAGAGCTATCGCTAGCGACGAAGTCGGCTGACGGCAACATGAACTCTGGAACGCTACATCTACCGCGACCATCAGATAGAAACATGCTCCACCAGCGACGCTAGCA AGAGACTCCGTCCAAACTCTGGACTGATGTCTCGTCCCACAGCACCCGGGAATGAAGACGTCTATACCCCTTAAGAACGATCCGCTCAGATACACCGGGTGAGTTATAGATTAGACAAGAATGTAGGCGGGCTCCATTAGCAGACAATGGGTGTTA AGCGACTTCAAGAAGAAGGCGACTTCAGCATGGGATGCGAATAAAGTAGGAGGTGCGGGTTTGGAATACAATGATCCCATAATCTTCAGATATAGGTAGGCGATATTGACAAGACAGATGTATAGTGGGCGCAGAAGCTGTGCATCTGAAAACTTA CCTACACTTGGCCAGCGATCCTCAATTATTGGTCGATTTAAGCCGTTAGTCAGATACAATGTGTATTCTGTCACGCGCGGTTAAAACGGCGGGTCACGCACATAAGGTCTCGAGGCCCTTAGATTGGCGGCTAAGCAAAGCCTAAAGTTCCGATCC TGTGTACGCAATCATATCGGACCTAATTGGATGTACATCAGATATATGGCGAGCTTGGGCCGTTCTAGCAGTGGCTTTCTAAGGCAGTCGTACGGAGAATATGTGGTCAAGTTTGTCGACTGAATCGTAGTACTTGGCACGCAGCTTAGCCACAGC TGACTCCGTTAGCAATTGCTTAGCCTCAGATAAACTTGGCGTGATATCGGCACGGAACGTGCCCTGAGTGCTGTTGGGGCAGTGGATGTGAGTGCACCCGTTTTATGTCGTGAAATGCCTGGAGTGATCGTCGAAACGACGTGCTAAGTTGGCCGA TAAATTTGTGCAATGACGGGTGGTACAAAACTATTATATTTTCGCGAGGATCCAATGCCCTTCAGATAGACAGTCTTCATGAAAGTCCTAATAGGTGGACCCTAGGGCTGTATCGGTCGGTATTTCCAAGCTCACGTCACGAGTAGATACACTCAC GTTGTCATGTCATGGGTCAACGTGAAAGACGAGTACCGTAGCACGGTATAGTCTATGGCCTATAGTTTGTCAGTCTTTGACCCAGCTACTGGACGGAATAGATTAACTTGTCCGAGGCGAATCAGATAAATTTCAAGTGACCATGATGATGCTTTC TCAAGTTCGTAGATCGAGGGTTGCGGATCTCAGCTCCGACTTACGTAGAGAATAGATCTCTTGCGTTCAATGGAGAAAGGGCTACCCATAAGCCATACTAATAACCCTATGGAGTGACGAGTCAGATACAGCTGTTAGCGGTACTGCGTAGACTCG TGCCCACCACATGTCGTGAAGCAGACTTGCATGCACGACATGAACTTAAACGCGGCATATCACAAGGTACTAGTGCCCTCTGCATTTATGTAGCGTAACCAATGGAAAATACGAATAGGCTGTGTACAGTTTTTCAGATACACAGTCAAGCTGAGG GTGACTTTACTGCACGTTGCTTCCAGGAACTGCCGACTTAATCGTGAGCGAAAACTAGTTCTCAGATATAACAAGTTCGAATAAGAATTTCACAGACCGGAATCGGAGGGATAGCCATAAATCTGGGTATGATGGATAGGTCCTGTCAATATTCGA TTTTGTAATGAGTTACGCTTTGCTTGGCTAGCCTTTGACCAAGTATATCCCATTTGCCTCCAACTCCGCATGACTCTGCCGGCGTCTACAGTATCAGTCGATAACAGGGGCCGCCCATCCAAACGTTCAACAGTCTTTTCTCAGCTCAGATACAGA GCGATGTGCGTTGATCGAAGGTAAAGTCCCGCGACCCACATTGCCGCTAATCTTCATTTGATCGTTGTAGAGACGTCCCGCGTAGAATCATTATCGATCAGATAGAACGCTTTAACGACCACCACTACTCGCTAGCCTTAACTTACGTTTGTCTGT GAAGTTTAATATCGAAACTTGGCTACATTGCCGTATAAGAGCTTACCGGCCCTACAAAGGATAGGAGTCAGCGCGGATCCAAATATCAGATAAAACCATCCACACTAGGATAGCAAGGCCATGCGCGGTGCCCTCAAAGAATAAGTGTGGAAGGGA CATTCCCTGACAATACGATGCACGGCTACAGTCTAGGCACTGGATACCTACGCGGCTGGCGTATCAGCAGCGTGGTTACATTAGACTTCCGGATGACGTCTTTAGGAAATCAGATAGATTATGGTTATCACTGGGCAACGCATCTCTATTGCTTCG ATGATTTTTCCGCAGTTTGCTTCTGGTTTACGTTTCCGCGGTTCTAGCTTCAGATAGAGTAAATTTCGGGGTCGGAGGTCCTCCACGATATTTGCGAACTCTTCGGATAGACATTTAGGCGCAAGCTCCGGCTCGCTATTGAAATCTTAGCGGGAG TTTTTGGATACGTGCCTAGCTTCCGCCGCAAACCATTCATCAAAAGCGTTCGAGGGAGGTTTGACTTCTGCTTTCAGATAAAGACGTCGCTATGCTCCCGAAATTAGCCTCGCGTATTTGCCCATATGAGCTAAACCGAACCTTGGAACTTTGTTT GGTAGGTAACCGAACATAGTAACATTACATGTCAGCTGGTTCCAAAGGGAGTCTGACTCCCCAGCGTTCTATGAAAGGGTACGCCTCAGATAAAGGGTCCGAGTTTTCACGGGTTGCGTACGGACATGGTCGGGGTATTAGGTGTACAATTACACC GCCAGCCTGAAACAGTTCTGGGGCCTCAGATACAATCCGCTTTCAGTGAAGCGCCCGGGTCGCTCTACTGTCAGACTCGAGTCTCACATATTTCCCCCAGAACGGCAGTCATGGGTCAGACATGGGAAGCAACAAGAGCTCTGTCACACTACGGCG AAGCCTCATTATCGGAATGCAGTATGGTACTCCCGCGTCAGATAAAGGTGCGGCCATCAATCTCTCTTGGTCTTAACACTCGAACTTAAATTACAAATCTATTATTTCGCTGTAGTAGCCCTTCGGCATACTCGATATCGCCAAATCTACGCTTCT GCCGAGCGCAACGTCTTTGCAGCGCATGTGTTAGTGACGTGACGTATGTTCAGATAGATTCGTTTACAACAGAGTTAATTGTCCAGGATAACACTCCATGAACCTGTACAATTTCGACTCTCACACGTCCGTGCCCAATGGAAAACCCTTACGGCG GGCGCTTCACTACAATTTAACTTAGACGAAAGCGCGTGCCACGGTACACTTACTGAGCCAACTGCGCGTGCCGAGGAGAGGAATACGTCTGCTTTCGTCAGATACACCGTAACCGTCTTCAGTCGATATCCGTCCAGTGTAGTCGATTGCTGCGCA'\n",
    "\n",
    "print('Random = ', ' '.join(str(e) for e in GreedyMotifSearch(dnas, k, t)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AACGGCGACACG TAGAGAGATACG GGCACCCCAGAA GTTCACACACCC CCGTGTGAGAAT ATCAGAACCGGG CGAGCCGGTGAG GAGTGTTCTGTC TCCGTATCCCCC ATAATGTCGGTT GCATGGCATGCC GCCGGGGTAAAC ACTACGTTGAAA ACGGTGGTTGCA CCGCATCGCTGT TGGAATCCCGCG ACCCTACTCGCT TATTCTGGATAA ATGTTTCAGCCG ATTAAGATTCAC AAGATGCAACGG ACTTCAGTTAGG AACTGCTGGGGC GGCTGATGTGCA TTTGTAGCAGCG\n"
     ]
    }
   ],
   "source": [
    "# Random Test\n",
    "k    = 12\n",
    "t    = 25\n",
    "dnas = 'AACGGCGACACGGTCCGACTGCACAACCAACTATCGCTAATGTTCCTCGCTACTTTAAAGTAATTGGGTCGCCATTGAATACAGCTCATCTATGCGATCTACCCTCGGTGAGTGGTGCATAATACGTGAAGTATTAAATGTTCTGCTTTAGCAGCA TAGAGAGATACGACCTGCGGAGAGGCTCTGACATAATTCTCGAATCCCCATGTCCTACGTGTTCATGACTATCAGCGCGGCCCCCTCTACTCTGGGAAGCCCGTGTACTGGTAAAACACATCATCCCTATCCACGAGCTGATTAGAGGGCGGGCAA GGCACCCCAGAAGGATCTGAGACACAACTCTCATCTCTATCCATTTGAGGGGCGTAAATCGGCAAGTATACACACCCTCGGGTTGTACCCCAACTAACTCTCAGCATATTGCAATGCCTCATCTACTCTTTGCCCCTACCATTTATGGGACTGGAC GTTCACACACCCAGAGGCCAATTTATGAGATAAGGACACAGCAGATACGTGATTTCATTCCCCTGAGGGTCGCGCTACCAAGGCGCATTAGCCTCATAGAAGCGGAGGCTACCTGGGAACTAATTATGGGTAATCTACTCTTGGCCTAAAGCCATA CCGTGTGAGAATCGAGTAGATCGATAACAGGTAAACCCTGGCTGCTAGTTCTACGCTTCGCCTTTTCCAGCGACAGGGAGATGTTCACTTAATCTGGGCCTCACGAAAGGTAGAGCCCCGCACGGGACGCCTAACTATCTGAATGTATGAAAGTTG ATCAGAACCGGGCTCTACGCTGGGTATGACTTTTCAAGTCGATTCGGCGTTAAAAACGCGCGGGCCTCCTGCTTGTAATTTGGTTGTTGGTGTATCGGCCTCTAAATGTCAGTGCGTATAAGAGGCAACCATGGTCTAACGCTCTATACCCATGTC CGAGCCGGTGAGGGCTTGTCGCAGCAACTATTAGTGATTTAAGGTGCGCGACAGCGTTCTCGGCGTATTGTCGTCTCCCGGTACGTGGGCGGTCAACAATATATATGACTCTACTCTAGGAGATGTTGGCGGCGAAATTGATGGTGCTAGGCAAAG GAGTGTTCTGTCGAAGCAGGTTCATCCGCCATTGGGTTCTACGCTCTGTCTCATCACTGGGCTAGAGCTGTCTGATACTATTCGCCCTTGTTGAAGTCGGTGAATACCTGGAGTCGGGATTACCAACTCTGAACGCATATTATGACGTCGAAACCG TCCGTATCCCCCCAGGGCCCGCGATATTTAGCTTGTCCCCCCTTGCCCTTCACACGGAATGTCGAGGTTGTATATGATGAGCAGATCTACGCTTTGCGCATAGCCCGTGGTGTGGATTTTCCTTCTGCCTTTGGATCTGAAGCCGTCCGCGCTCCA ATAATGTCGGTTAGGCTACAGCTAGATTATTTGCCTTGTGTATACCCACGACACCATCGTCGATATCTTGCAACTGTGCTCATTCCAAACCCTTGTCAATTCCCGACACAGAAAAGGGCTTTCTACACTAAGTTAGAACTTTAGAAAGGTACGTAG GCATGGCATGCCCGTCGACCGGTGCTCTACACTTAGGATACTTAGTTCTGTGTAATGTCTATGCTTATCCCAGATTAAGCCCATCAGGGCACCGTAAGCGTAAGCCAGATGTTGAGCCTGGGCCATTCCGAGCTTAAGGTCATTGATTGCAATTAG GCCGGGGTAAACTGCGCGTGGTTGAGGCAGCCCATCACGCTTTATCGGCTGTTCGACCGTTAGTCATCGTTACTCTACCCTAAGCCGGAGAATGGGCAGACATGTCTACCCAAGGCGGCAAACGACGATGTGTACGTACGACCCGCAGAGGCCCGA ACTACGTTGAAAGTCTACACTCTGGCAGGTCATGATCTAAAGCGAGTCATTGTGCTCTTGCGTTGCAGGAGACCTCTATCAGTATGTCACATAACGTCCAATCATTACGTCGGCATCTTTTATTGCGCAGTTTTCAGGCCCCATGTAAGACTGTCG ACGGTGGTTGCACGGGTTTTCAGGCCATGCAAACACGCCGTCAACGTCCATGACGAGTTTATCTACTCTATGTCGGGCATCCTTACACTTATACAGATGTGCTCCTGACCCATTTACACAACTCCGTTCGCAGAATCCACACGAATGAATAAGCAT CCGCATCGCTGTATCCGATCGGAACTCTACTCTGGGTGGAGGGCTCTCTCCTTAACGAATGACATGCTTGATCATTAAAGTGCAATCGACGAGCGAATCTCGCACCCTTTATCCAACAATCGAAGCCTGTTACGGACTAGCGGGTCCGAATATGGA TGGAATCCCGCGTTCTACCCTGTGGTAAAAGGGCATGCATGGCTCCCGTAGGCTACTCGTGAAGGCTGACTAAGGAACGGCGGTATGCAGGCCAAACTGGCGGTCCATATCGTAGGGGGGCCTCGACCCGTAATATTACATGCCGCTATTCCGGCA ACCCTACTCGCTTCGAGAATACTCGTCTACCCTCAGGGTAGTTCTAGCCAGAACAATACAGGCTTCAGCAGAGGCTTTGCCGCGACTCTCTAGGGGTCATCACCGCTTGTGCCGGCTCCAATCTGATCTGAGGAGCCGAGGCGGAGGTCTTCGGTT TATTCTGGATAAATCTACTCTTTGTCTGCAGGAGGCATCAGCCTCCAAACAGCTATATTAGTCCAGTACTCTTGTCCTACTTAATCGACTGTGGCATAAGGTTTGTTCGCACATAAACCGGGAGTTCTGATCAACTGTGGGTTCTCCTGTCGATGG ATGTTTCAGCCGGATTTCTAGCCGTAACGCTAGTTTTCCGATGAGCAAAGTGTAACCCAAGATGTAGTTACCCTCCGATCAAGCCAATGAGCATTGATCTACCCTGCGGATACTCCGGGAACGTATTAGTAGCAGCGGCCTCTTTGACCATCTCCG ATTAAGATTCACGGTGGCAGCAAAACACCATAGCAGGATTCCATCTTATTGCTCGAAGGACTCTACCCTAGGATGCCCAATAGAACTACCAACTTTGGGGTATATGGATGGGTCCTAGCAACGCGCCCCACGTGGCCCTGGAATAACCGAATGTAG AAGATGCAACGGTCGACCAATTGTGTCTACTCTAAGAACTGTGAGGCCACTAGTTCATGAGCATTAATCGTCGCATATATTCGCGCCGTGATGGAATAAAGTCGTGGCATCCAGTAGAACACCGAAAGTTTTTTCTTTTAGACTCCTCGTCCGAGG ACTTCAGTTAGGCGGAGCCAAGGAACCACCTAGCACCATACACACCCCAGGTCCGTAGCGCAATGTATGAAAGCATAGTTAGGGGATTTAAACGCGGCTTAACTCCTAGTCTACCCTCGGGCATTACCCTTGCTTCAAGGGTCGACCCAGTGCCAG AACTGCTGGGGCATCTTGCTGTTTAGTCAGAGCGCCTTTCTATATATGCCGCGACAAATCCTACTCGTGCCCAACCGGAAACGTCCGCCGCACCTAACGCCGCGCTCTGGGCTACCACACAGATCTTCTTCCCGTATCGATGATCTCTACTCTTAG GGCTGATGTGCAATGAGGTATTGCCTCTGAGGTGCGATCTACGCTGAGATTCGAGTGACATGGCTAGCGCCCAGCGTTGCAGGCGATACGGACAGTCACCGCCGGTTCGGAGTATTGACGGGGGAGTGAGAATTGCGAGTATAACTCGACGGGAGA TTTGTAGCAGCGTTAACTATGGGCGCGAACCTTATCGTCTACCCTCGGGCAGTTCGGCGACCCATATCTCAACACCTTATTAATGTTAATCCTAGATTTGGGCATGACAATTGCCTCGTATTAGTGGCAGCGGCACCAATCCCACAAGCCGGTTCT'\n",
    "sequences = dnas.split(\" \")\n",
    "\n",
    "print(' '.join(str(e) for e in GreedyMotifSearch(sequences, k, t)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample should be (TTC ATC TTC ATC TTC)? TTC ATC TTC ATC TTC\n"
     ]
    }
   ],
   "source": [
    "# Stepik 1.6 Greedy Motif Search Algorithm with Pseudocounts\n",
    "\n",
    "# Input:  String dnas, integer k (kmer length) and integer t (sequences)\n",
    "# Output: matrix profile\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def count_occurrences(motifs):\n",
    "    counts = {'A': [], 'C': [], 'G': [], 'T': []}\n",
    "    k = len(motifs[0])\n",
    "\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        counts['A'].append(column.count('A'))\n",
    "        counts['C'].append(column.count('C'))\n",
    "        counts['G'].append(column.count('G'))\n",
    "        counts['T'].append(column.count('T'))\n",
    "\n",
    "    return counts\n",
    "\n",
    "def ProfileMatrix(motifs):\n",
    "    counts = count_occurrences(motifs)\n",
    "    k = len(motifs[0])\n",
    "    profile = {'A': [], 'C': [], 'G': [], 'T': []}\n",
    "\n",
    "    for nucleotide in ['A', 'C', 'G', 'T']:\n",
    "        for i in range(k):\n",
    "            profile[nucleotide].append((counts[nucleotide][i] + 1) / (len(motifs) + 4))  # Adding pseudocounts\n",
    "\n",
    "    return profile\n",
    "\n",
    "def GreedyMotifPseudocounts(dnas, k, t):\n",
    "    if type(dnas) == str:\n",
    "        dnas = dnas.split(\" \")\n",
    "\n",
    "    best_motifs = []\n",
    "    best_score = float('inf')\n",
    "\n",
    "    for string in dnas:\n",
    "        best_motifs.append (string[0:k])\n",
    "    base_string   = dnas[0]\n",
    "    other_strings = dnas[1:]\n",
    "\n",
    "    for i in range(len(dnas[0]) - k + 1):\n",
    "        motifs = [dnas[0][i:i+k]]\n",
    "\n",
    "        for j in range(1,t):\n",
    "            profile = ProfileMatrix(motifs)\n",
    "            most_probable_kmer = ProfileMostProbableKmer(dnas[j], k, profile)\n",
    "            motifs.append(most_probable_kmer)\n",
    "        \n",
    "        # best_score    = DistanceBetweenStrings(best_motifs[0], best_motifs[1:])\n",
    "        # current_score = DistanceBetweenStrings(motifs[0], motifs[1:])\n",
    "\n",
    "        best_score    = Score(best_motifs)\n",
    "        current_score = Score(motifs)\n",
    "\n",
    "        #print(motifs, current_score)\n",
    "        \n",
    "        if best_score > current_score:\n",
    "            best_motifs = motifs\n",
    "\n",
    "    return best_motifs\n",
    "\n",
    "# Sample Test\n",
    "k    = 3 \n",
    "t    = 5\n",
    "dnas = 'GGCGTTCAGGCA AAGAATCAGTCA CAAGGAGTTCGC CACGTCAATCAC CAATAATATTCG'\n",
    "\n",
    "print('Sample should be (TTC ATC TTC ATC TTC)?', ' '.join(str(e) for e in GreedyMotifPseudocounts(dnas, k, t)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Random = GCTAATTTATAA GCCAATTCATCA TCCAATTGATCA TCCAATTCATGA GCCAATTTATTA GCAAATTGATCA CCGAATTCATCA ACAAATTAATTA TCAAATTCATTA CCCAATTTATTA TCAAATTGATAA TCAAATTGATGA GCCAATTGATAA ACTAATTCATTA TCGAATTGATTA ACGAATTTATCA ACAAATTGATGA GCGAATTCATAA ACCAATTAATTA TCTAATTTATAA ACTAATTTATGA GCTAATTTATAA ACGAATTCATGA TCCAATTTATTA ACCAATTAATTA\n"
     ]
    }
   ],
   "source": [
    "# Random Test\n",
    "k    = 12 \n",
    "t    = 25\n",
    "dnas = 'AGATAACCGCCCAATTTGGCTTAACAGAGCTGAATCATACTAATACAAGTGTATAGCATGACTTCTCGCTATCACCGTTAGGTATAGGTCCCCCTGGCCGCGAGACACCTAGGCAGGTAGGCTAATTTATAACGACCAAGCCTTAACTGCGGTCAG ACGGAGTGCGTGTCTCCCTGCTCGGACAGTGACTACAGGGAGGAGGCTCTATCCATGGGTGCCAATTCATCACAAATGTCTATAAATATCGTGAGGTAATGGGATCGAATTGCCGAAAGCTCCATACATAAGCAGTTACCCCTTCGGCCGGGTCGG GGCCAGTCTGGTACACGTGCTCCGAAGAGCTCTGATCGATCAGCAATCCATACGGTCAACCCAGCTTATACTCGCCATCATGTGTCCAATTGATCACTGAGCAGTTGCGCTGTTTATAAATCCGGCCATTCGGAAATAGCGAGTCAATAATCTACT GACCCAACTAATGGTAGGCCGATTAAGTGATGGTTAGAGAGTATATCGCGTGATCTACTGCGTCAATTGTAGCTTTTACTTTCGTCCAATTCATGAGTTGGTCATGCGCTGCGCGTGAGCTAGTGATCGGGGGAGGGTCTAGGACGATCAAGTCAT GCGGATCAAATTAGCTCATTTGCGTCGTCCTCTTGTCGGGTTCCCCCCCGGAGATGAGACAGATAGCTAGCCGCCAATTTATTAGCTTTCACTTGTCACTCGGAAGGGGGTGGCTCGAACAAAGGACTGACTGCCCCTGGGAATTAACCGTTTATT GCGACGCAATGAGAAACGGATGGCTAGAGAGGAGCTAAATTGTGCGTCACGCGACTAGTCGCAAATTGATCAAGAAAGATTCCTACTTATAGGGAGTAAGAGTTAACGAGTAGCAGGGCTAACCTGCTATCTTACTACTTCACGGGCACTCGTCTA GAGTGCGAAGTTAAGGCGTAATGCAACTCTTTAACTTTGAGCGCTGCGACTGGCCCTTAATCGACACATGGTCAGAACTGCCACACGGTCCTCGCCGCTGCGAACAATCAGTCCAAGTCACACCACAAGACTTCCACGCCGCGGCCGAATTCATCA ATTTAGGCCAAACCCTGGCATATACGGGTGTGCTTAAAAGTCAAGGGCACAAATTAATTAGGCATTGAGCAGCAAGCCGATACCAAAGTTCGCTCACGCACGGCAACTATCCCTTCCTAAGTACGACCACGCCCCATGTATAAGTCCCCCGATCGT TATTTCCAGTCATAGTACGAACCGCGCACTAGCTAGTGGTCCTTAGCAGTGTCGATCTTATCGTATGAGAAGTATGACCGTACGTAATAGCGGATGGTCGGCTGATAATTCTGCTAGACGTCAAATTCATTATTCTGGGGCGCCGGCATTATAACA TCGGCGACAATACTAGGGTTTGTAATGCTGACCAACTTTGTCTCCTTGGGGTCGCGGTTAACGCGCTGTATTCTATAAAAACTTTGACCTGGGACCTCTTTCCCTGGACCCAATTTATTAAGTCGGAGAACCACATGAAGAACAACAGAAATACTC GGTACCTCAGACGTAGAGATGAAGCAAATCAAGAGTTTTCTGACGGGACGACGTAGGTAAGAAACCGTCCAGCTAAAGATAAGGTCTTGGTCAAGAGCGCTGGAAACATGCCACCGAGGTTATTCTTAACCATCAAATTGATAATGCTACGTCTTC TCCTTAATATTCGGCTTTTCCCTTATTCTTGAGGTCTTAAAGGAATTCGCTAATGGCGACTCGTATGCGACGTAATAAATGCTTACGTGCCTCGATTCAAATTGATGAAAGTCCGGTGTCCAGTCAAATCCAGATATGGGAATGGAGCCAGCATGG GCCAATTGATAAGTAAAGATGCTGTACCACGTCCTCTAACCTCCAAGAGAATCGCCACTCTTGGAGGCACTTCATGTATATCACGGCTATTGGACTGATTTTTAGTAGGATGTTTCCGTTCGTCTTTGATTCAGGACCCTTTTTAGTCGCCTATGG ACATCTATCGTTACTGGACTAGGGAGGAAACATCAGTTTGGAGATACCTCTCTAAACGCGTACACACTAGCGTAGGTTTTTCAAGATACTCCAGTTTGCAGGGCTCTGGGGCCTGGTCACACTAATTCATTATGCGTGCATATAAGTGCGACAAAG TTCGATTGAATATCGAATTGATTAGTCCCCTAATCGAGTACCGGGTATACGATTAGACTGGTGTCGCTACTGTTACTAGCAGATCACATTTTGAGCCGCATGAGCATGAAATTCTTCCGAGCCCGCGACGGATGCAGCTCAAGCATGCAGGTAGGG GCGAACCCGATCAATTACTTCAGAAAGGTCGTTGGGGCGATCCTGCGACCAGTAGTAAGAGACTCTCGACTCTGCATAATCTCGCAGCTCTGGGTGACGTGACGATTCCGTATCGAAAATACGAATTTATCATGCCGTAGGCGAATCATGTTGCAT ATTTAGGAAACTTCCGTAGCGATCCCCTAATTGCTGCCTCATCTGAACGCTTCACGGACAATTGAGGAGAGGTGCACTTTGTCCGTCGGATTAGTCATGCTGTGGTCCCAACGAACGCCAAGAGCTCGGACTCAGTGGGAGGGTACAAATTGATGA GTCAAGAGCGTATCTTTACTATATACGCCTCATAGGCAAGCTATATTGGCGAATTCATAACTTCTAAGTCCGTAAGTAAACATCACGGGTTCGAGAGAAAAGGGAAGGTCGAATGGGGCAGGCTTGCGCCTCGATGTGTACCATTCCTATTTCATA TGTTTCAGAGGATCGAGTTAGCCCCCGAGTTTACATTGGGCGACAAGGATCAGCTTGTATACAAAGTGGATGTCGTCCGGAGTAAACAGGCTACGCCCCTGTGATCGCACCAATTAATTATAGAGGATCGTGTGAGCTTCACTCATGGTCCAAGCC AAACTCTTCTAAGTTATATTCCACGGGGATGGGTGATCTAATTTATAATAAATAGCAGGTGAGCATCCGCACCCTGGTACTTTATGCCTGGTTCGAACCCGCTGGCGTTACGTCTTCCCATAGTGCCTCTTTCGGCGCGCTCGCCCTGACTCTCGT TTATGTACCCTGATACTTCGACCGCTCCCCCCTGTAATTTTCACTTTGTGCGTACACTCAATAACATGTGTGAATTTACGTGTAGTGAGAGCATATCTGCGTACTGCCGACGGTCTTAGCTCCGTAGATCCGACTAATTTATGATCGTACTGAGTC GCTAATTTATAACATTAACAAGCGTAAATCAAATGTGAGCACGTTCTTTAGTTAGCAATGCCTTATTTTGTCCTAAGAACAGTAGGCTACAGTACCAGGGGGTATCCATACTGCAGCCATATGGTCAACATATAAGTGATATCATGATTTCTCAGC TTTATAGCCTATGCCGAAGTGTTGCTTTAGATCCATAGAACCTCTTCGCCTCTCAGGTTTACGAATTCATGAGTAAGTAGGCTTATCTCAGAGGTAGCGACCATGGTAACCCCCAAACGTTTTCTCCCAGCTTGTCGGGGAGTACCCCCCCCGGTT CGATGAGTACGCCCTCTTACATGAGATCGTGGAAATTTATTATGTGAAGCTGAACCCCTAAAGATCCCAAGGTCCAATTTATTACCTTACGGCAACTGCTCATTCTAAGGCTTGATCGGCCGCCGGACCCCTGTGTCTATGTGTGAATAAAGTTAA CTTCCCGCAAATACGTCGCTGTCCCACTCATACTTGGCGGCTCCGCGGACCAATTAATTATCGAGCCACCGTAGGCTCTTCCATTCAGGTAGCGCCACCGAGCTTGCTCGCTTTCCCACGACGCTTATGCGCTATGTACTTCGCAGTTGACAACAC'\n",
    "\n",
    "print(\"Random =\", ' '.join(str(e) for e in GreedyMotifPseudocounts(dnas, k, t)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entropy A: 1.0\n",
      "Entropy B: 2.0\n",
      "Entropy C: 0.0\n",
      "Entropy D: 1.5\n",
      "Entropy distribution from lowest to highest\n",
      "Distribution C: Entropy = 0.0\n",
      "Distribution A: Entropy = 1.0\n",
      "Distribution D: Entropy = 1.5\n",
      "Distribution B: Entropy = 2.0\n"
     ]
    }
   ],
   "source": [
    "# Week 3 Quiz - Question 3\n",
    "\n",
    "# Order the following probability distributions from lowest to highest entropy:\n",
    "# A: (0.5, 0, 0, 0.5)\n",
    "# B: (0.25, 0.25, 0.25, 0.25)\n",
    "# C: (0, 0, 0, 1)\n",
    "# D: (0.25, 0, 0.5, 0.25)\n",
    "\n",
    "import math\n",
    "\n",
    "def calculate_entropy(probabilities):\n",
    "    entropy = 0\n",
    "    for p in probabilities:\n",
    "        if p != 0:\n",
    "            entropy -= p * math.log2(p)\n",
    "    return entropy\n",
    "\n",
    "# Example usage:\n",
    "A = [0.5, 0, 0, 0.5]\n",
    "B = [0.25, 0.25, 0.25, 0.25]\n",
    "C = [0, 0, 0, 1]\n",
    "D = [0.25, 0, 0.5, 0.25]\n",
    "\n",
    "entropy_A = calculate_entropy(A)\n",
    "entropy_B = calculate_entropy(B)\n",
    "entropy_C = calculate_entropy(C)\n",
    "entropy_D = calculate_entropy(D)\n",
    "\n",
    "print(\"Entropy A:\", entropy_A)\n",
    "print(\"Entropy B:\", entropy_B)\n",
    "print(\"Entropy C:\", entropy_C)\n",
    "print(\"Entropy D:\", entropy_D)\n",
    "\n",
    "# Or define the probability as distributions\n",
    "distributions = {\n",
    "    'A': [0.5, 0, 0, 0.5],\n",
    "    'B': [0.25, 0.25, 0.25, 0.25],\n",
    "    'C': [0, 0, 0, 1],\n",
    "    'D': [0.25, 0, 0.5, 0.25]\n",
    "}\n",
    "\n",
    "# Calculate the entropy for each distribution\n",
    "entropy_values = {}\n",
    "for distribution, probabilities in distributions.items():\n",
    "    entropy = calculate_entropy(probabilities)\n",
    "    entropy_values[distribution] = entropy\n",
    "\n",
    "# Sort the distributions based on entropy values\n",
    "sorted_distributions = sorted(entropy_values.items(), key=lambda x: x[1])\n",
    "\n",
    "# Display the distributions in order from lowest to highest entropy\n",
    "print(\"Entropy distribution from lowest to highest\")\n",
    "for distribution, entropy in sorted_distributions:\n",
    "    print(f\"Distribution {distribution}: Entropy = {entropy}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Consensus String Options by Position = ['A', 'ACG', 'G', 'CT', 'GT', 'A']\n",
      "Matching options: ['AGGCTA', 'ACGTTA', 'ACGCGA']\n"
     ]
    }
   ],
   "source": [
    "# Week 3 Quiz - Question 4\n",
    "\n",
    "# Consider the following profile matrix:\n",
    "\n",
    "'''A:  0.4  0.3  0.0  0.1  0.0  0.9\n",
    "C:  0.2  0.3  0.0  0.4  0.0  0.1\n",
    "G:  0.1  0.3  1.0  0.1  0.5  0.0\n",
    "T:  0.3  0.1  0.0  0.4  0.5  0.0'''\n",
    "\n",
    "# Which of the following strings is a consensus string for this profile matrix?  \n",
    "# (Select all that apply.)\n",
    "\n",
    "def Consensus(profile):\n",
    "    k = len(profile[0])\n",
    "    consensus_list = []\n",
    "    for j in range(k):\n",
    "        m = 0\n",
    "        frequentSymbols = []\n",
    "        for i, symbol in enumerate(\"ACGT\"):\n",
    "            if profile[i][j] > m:\n",
    "                m = profile[i][j]\n",
    "                frequentSymbols = [symbol]\n",
    "            elif profile[i][j] == m:\n",
    "                frequentSymbols.append(symbol)\n",
    "        consensus_list.append(''.join(frequentSymbols))\n",
    "    return consensus_list\n",
    "\n",
    "\n",
    "profile = [\n",
    "    [0.4, 0.3, 0.0, 0.1, 0.0, 0.9],\n",
    "    [0.2, 0.3, 0.0, 0.4, 0.0, 0.1],\n",
    "    [0.1, 0.3, 1.0, 0.1, 0.5, 0.0],\n",
    "    [0.3, 0.1, 0.0, 0.4, 0.5, 0.0]\n",
    "]\n",
    "\n",
    "print('Consensus String Options by Position =', Consensus(profile))\n",
    "\n",
    "# Returns AAGCGA, which is not an option for quiz.\n",
    "# I believe AAGTGA, ACGCGA and ACGTTA are valid consensus strings\n",
    "\n",
    "'''options = [\n",
    "    \"AAGTGA\",\n",
    "    \"ACGCGA\",\n",
    "    \"ACGTTT\",\n",
    "    \"ATGCTA\",\n",
    "    \"ACGTTA\",\n",
    "    \"AGGTCA\"\n",
    "]'''\n",
    "\n",
    "options = [\n",
    "    \"TCGCGA\",\n",
    "    \"AAGCCA\",\n",
    "    \"AGGCTA\",\n",
    "    \"ACGTTA\",\n",
    "    \"AGGTCA\",\n",
    "    \"ACGCGA\"\n",
    "]\n",
    "\n",
    "matching_options = []\n",
    "consensus_list = Consensus(profile)\n",
    "\n",
    "for option in options:\n",
    "    is_matching = True\n",
    "    for consensus_chars, option_char in zip(consensus_list, option):\n",
    "        if option_char not in consensus_chars:\n",
    "            is_matching = False\n",
    "            break\n",
    "    if is_matching:\n",
    "        matching_options.append(option)\n",
    "\n",
    "print('Matching options:', matching_options)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Question 5 Median String = ['GTAGGAA', 'TAGTTTC', 'GAACCAC', 'AATCCTA']\n"
     ]
    }
   ],
   "source": [
    "# Week 3 Quiz - Question 5\n",
    "\n",
    "# Consider the following motif matrix:\n",
    "\n",
    "# CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC \n",
    "# GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC \n",
    "# GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG\n",
    "\n",
    "# Which of the following 7-mers is a median string for this motif matrix? (Select all that apply.)\n",
    "\n",
    "def MedianString(genome, k):\n",
    "    distance = float('inf')\n",
    "    median = []\n",
    "\n",
    "    # Iterate through all possible k-mers\n",
    "    for i in range(len(genome[0]) - k + 1):\n",
    "        pattern = genome[0][i:i+k]\n",
    "        cur_distance = 0\n",
    "\n",
    "        # Calculate the total Hamming distance for the current pattern\n",
    "        for j in range(len(genome)):\n",
    "            seq = genome[j]\n",
    "            min_distance = float('inf')\n",
    "\n",
    "            # Calculate the minimum Hamming distance to the current sequence\n",
    "            for l in range(len(seq) - k + 1):\n",
    "                kmer = seq[l:l+k]\n",
    "                hamming_dist = sum(base1 != base2 for base1, base2 in zip(pattern, kmer))\n",
    "                min_distance = min(min_distance, hamming_dist)\n",
    "\n",
    "            cur_distance += min_distance\n",
    "\n",
    "        if cur_distance < distance:\n",
    "            distance = cur_distance\n",
    "            median = [pattern]\n",
    "        elif cur_distance == distance:\n",
    "            median.append(pattern)\n",
    "\n",
    "    return median\n",
    "\n",
    "\n",
    "k    = 7\n",
    "# genome = 'CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG'\n",
    "genome = [\n",
    "    'CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC', \n",
    "    'GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC', \n",
    "    'GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG'\n",
    "]\n",
    "\n",
    "print(\"Question 5 Median String =\", MedianString(genome, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute Pr = 0.0024000000000000002\n",
      "Compute Pr = 0.0054\n"
     ]
    }
   ],
   "source": [
    "# Week 3 Quiz - Question 6\n",
    "\n",
    "# Consider the following profile matrix Profile:\n",
    "\n",
    "# A:  0.4  0.3  0.0  0.1  0.0  0.9\n",
    "# C:  0.2  0.3  0.0  0.4  0.0  0.1\n",
    "# G:  0.1  0.3  1.0  0.1  0.5  0.0\n",
    "# T:  0.3  0.1  0.0  0.4  0.5  0.0\n",
    "\n",
    "# Compute Pr(AAGTTC|Profile). (Express your answer as a decimal and do not round your answer.)\n",
    "\n",
    "profile = {\n",
    "    'A': [0.4, 0.3, 0.0, 0.1, 0.0, 0.9],\n",
    "    'C': [0.2, 0.3, 0.0, 0.4, 0.0, 0.1],\n",
    "    'G': [0.1, 0.3, 1.0, 0.1, 0.5, 0.0],\n",
    "    'T': [0.3, 0.1, 0.0, 0.4, 0.5, 0.0]\n",
    "}\n",
    "\n",
    "print('Compute Pr =', ComputePr('AAGTTC', profile))\n",
    "\n",
    "print('Compute Pr =', ComputePr('GAGCTA', profile))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
